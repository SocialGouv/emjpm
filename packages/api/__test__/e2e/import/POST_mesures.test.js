//

const request = require("supertest");

const { knex } = global;
jest.setMock("@emjpm/api/db/knex", knex);

const server = require("@emjpm/api/app");

const { getTokenByUserType } = require("../utils");

//

beforeAll(async () => {
  await knex.migrate.latest();
  await knex.seed.run();
});

afterEach(async () => {
  // Remove all created mesures
  await knex("mesures")
    .whereNotBetween("id", [1, 5])
    .delete();
});

// strip the created_at date in the response
const simpler = ({ created_at, ...props }) => props;

test("should import mesures correctly", async () => {
  const token = await getTokenByUserType("mandataire");
  const response = await request(server)
    .post("/api/v1/mandataires/mesures/bulk")
    .set("Authorization", "Bearer " + token)
    .send({ sheetData: [1, 2] });
  expect(response.body).toMatchSnapshot({
    message: "2 Mesures importÃ©es."
  });
  expect(response.status).toBe(200);

  const mesures = await knex
    .from("mesures")
    .orderBy("id", "desc")
    .limit(2);

  expect(mesures.map(simpler)).toMatchSnapshot();
});

test("should skip existing references", async () => {
  const token = await getTokenByUserType("mandataire");

  const curMesures = await knex.from("mesures").where({ mandataire_id: 1 });
  expect(curMesures.length).toBe(4);

  {
    const response = await request(server)
      .post("/api/v1/mandataires/mesures/bulk")
      .set("Authorization", "Bearer " + token)
      .send({ sheetData: [{ numero_dossier: "abc" }] });

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();

    const newMesures = await knex.from("mesures").where({ mandataire_id: 1 });
    expect(newMesures.length).toBe(5);
  }

  {
    const response = await request(server)
      .post("/api/v1/mandataires/mesures/bulk")
      .set("Authorization", "Bearer " + token)
      .send({ sheetData: [{ numero_dossier: "abc" }, { numero_dossier: "xyz" }]});

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();

    const newMesures = await knex.from("mesures").where({ mandataire_id: 1 });
    expect(newMesures.length).toBe(6);
  }
});

//
