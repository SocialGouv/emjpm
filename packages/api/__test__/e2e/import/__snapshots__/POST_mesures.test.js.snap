// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should import mesures correctly 1`] = `
Object {
  "added": 2,
  "message": "2 Mesures import√©es.",
  "success": true,
}
`;

exports[`should skip existing references 1`] = `
Object {
  "added": 1,
  "message": Any<String>,
  "success": true,
}
`;
